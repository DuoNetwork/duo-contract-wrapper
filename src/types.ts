import DualClassCustodianWapper from './DualClassCustodianWrapper';

export interface IEvent {
	contractAddress: string;
	type: string;
	id: string;
	blockHash: string;
	blockNumber: number;
	transactionHash: string;
	logStatus: string;
	parameters: { [key: string]: string };
	timestamp: number;
}

export interface IContractPrice {
	price: number;
	timestamp: number;
}

export interface IManagedAddresses {
	operator: string;
	roleManager: string;
}

export interface ICustodianAddresses extends IManagedAddresses {
	feeCollector: string;
	oracle: string;
	aToken: string;
	bToken: string;
}

export interface IManagedStates {
	lastOperationTime: number;
	operationCoolDown: number;
}

export interface IEsplanadeStates {
	isStarted: boolean;
	votingStage: string;
	poolAddrsHot: string[];
	poolAddrsCold: string[];
	custodianContractAddrs: string[];
	otherContractAddrs: string[];
	operationCoolDown: number;
	lastOperationTime: number;
	votingData: IVotingData;
}

export interface IEsplanadeAddresses {
	moderator: string;
	candidate: string;
}

export interface ICustodianStates extends IManagedStates {
	state: string;
	minBalance: number;
	totalSupplyA: number;
	totalSupplyB: number;
	ethCollateral: number;
	navA: number;
	navB: number;
	lastPrice: number;
	lastPriceTime: number;
	resetPrice: number;
	resetPriceTime: number;
	createCommRate: number;
	redeemCommRate: number;
	period: number;
	maturity: number;
	preResetWaitingBlocks: number;
	priceFetchCoolDown: number;
	nextResetAddrIndex: number;
	totalUsers: number;
	feeBalance: number;
}
export interface ICustodianWrappers {
	[type: string]: {
		[tenor: string]: DualClassCustodianWapper
	}
}

export interface IDualClassStates extends ICustodianStates {
	resetState: string;
	alpha: number;
	beta: number;
	periodCoupon: number;
	limitPeriodic: number;
	limitUpper: number;
	limitLower: number;
	iterationGasThreshold: number;
}

export interface IVotingData {
	started: number;
	votedFor: number;
	votedAgainst: number;
	totalVoters: number;
}

export interface ICustodianContractAddress {
	custodian: IContractAddress;
	aToken: IContractAddress;
	bToken: IContractAddress;
}

export interface IContractAddress {
	code: string;
	address: string;
}

export interface ICustodianContractAddresses {
	[tenor: string]: ICustodianContractAddress,
}

export interface IContractAddresses {
	Custodians: {
		[type: string]: ICustodianContractAddresses
	};
	MultiSigManagers: IContractAddress[];
	Oracles: IContractAddress[];
}
