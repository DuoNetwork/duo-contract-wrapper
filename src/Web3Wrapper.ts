import * as CST from './constants';
import { kovan, mainnet } from './contractAddresses';
import erc20Abi from './static/ERC20.json';
import { IContractAddresses, IEvent, ITransactionOption, Wallet } from './types';

const Web3 = require('web3');
const HDWalletProvider = require('./external/HDWalletProvider');
const BigNumber = require('bignumber.js');
const ProviderEngine = require('web3-provider-engine');
const FetchSubprovider = require('web3-provider-engine/subproviders/fetch');
const createLedgerSubprovider = require('@ledgerhq/web3-subprovider').default;
const TransportU2F = require('@ledgerhq/hw-transport-u2f').default;

export class Web3Wrapper {
	public web3: any;
	public wallet: Wallet = Wallet.None;
	public accountIndex: number = 0;
	private live: boolean;
	private providerUrl: string;
	public readonly contractAddresses: IContractAddresses;
	public readonly inceptionBlockNumber: number;
	public handleSwitchToMetaMask: Array<() => any>;
	public handleSwitchToLedger: Array<() => any>;

	constructor(window: any, providerUrl: string, privateKey: string, live: boolean) {
		this.live = live;
		this.contractAddresses = this.live ? mainnet : kovan;
		this.providerUrl = providerUrl;
		if (window && (window.ethereum || window.web3)) {
			this.web3 = new Web3(window.ethereum || window.web3.currentProvider);
			this.wallet = Wallet.MetaMask;
		} else if (!window && privateKey) {
			const hdWallet = new HDWalletProvider(privateKey, providerUrl);
			this.web3 = new Web3(hdWallet);
			this.wallet = Wallet.Local;
		} else {
			this.web3 = new Web3(providerUrl);
			this.wallet = Wallet.None;
		}
		this.handleSwitchToMetaMask = [];
		this.handleSwitchToLedger = [];
		this.inceptionBlockNumber = live ? CST.INCEPTION_BLK_MAIN : CST.INCEPTION_BLK_KOVAN;
	}

	public onSwitchToMetaMask(handleSwitchToMetaMask: () => any) {
		this.handleSwitchToMetaMask.push(handleSwitchToMetaMask);
	}

	public onSwitchToLedger(handleSwitchToLedger: () => any) {
		this.handleSwitchToLedger.push(handleSwitchToLedger);
	}

	public switchToMetaMask(window: any) {
		if (window && (window.ethereum || window.web3)) {
			this.web3 = new Web3(window.ethereum || window.web3.currentProvider);
			this.wallet = Wallet.MetaMask;
		} else {
			this.web3 = new Web3(this.providerUrl);
			this.wallet = Wallet.None;
		}
		this.accountIndex = 0;
		this.handleSwitchToMetaMask.forEach(h => {
			try {
				h();
			} catch (error) {
				console.log(error);
			}
		});
	}

	public async switchToLedger() {
		const engine = new ProviderEngine();
		const getTransport = () => TransportU2F.create();
		const networkId = this.live ? CST.ETH_MAINNET_ID : CST.ETH_KOVAN_ID;
		const rpcUrl = this.providerUrl;
		const ledger = createLedgerSubprovider(getTransport, {
			networkId,
			accountsLength: 20
		});
		engine.addProvider(ledger);
		engine.addProvider(new FetchSubprovider({ rpcUrl }));
		engine.start();
		const newWeb3 = new Web3(engine);
		const accounts = await newWeb3.eth.getAccounts();
		this.web3 = newWeb3;
		this.wallet = Wallet.Ledger;

		this.handleSwitchToLedger.forEach(h => {
			try {
				h();
			} catch (error) {
				console.log(error);
			}
		});
		return accounts;
	}

	public onWeb3AccountUpdate(onUpdate: (addr: string, network: number) => any) {
		if (this.wallet !== Wallet.MetaMask) return;

		const store = (this.web3.currentProvider as any).publicConfigStore;
		if (store)
			store.on('update', () => {
				if (
					this.wallet === Wallet.MetaMask &&
					store.getState().selectedAddress &&
					store.getState().networkVersion
				)
					onUpdate(
						store.getState().selectedAddress,
						Number(store.getState().networkVersion)
					);
			});
	}

	public isReadOnly() {
		return this.wallet === Wallet.None;
	}

	public readOnlyReject() {
		return Promise.reject('Read Only Mode');
	}

	public isLocal() {
		return this.wallet === Wallet.Local;
	}

	public isLive() {
		return this.live;
	}

	public wrongEnvReject() {
		return Promise.reject('Wrong Env');
	}

	public getGasPrice(): Promise<string> {
		return this.web3.eth.getGasPrice();
	}

	public async sendEther(
		from: string,
		to: string,
		value: number,
		option: ITransactionOption = {}
	): Promise<any> {
		const gasPrice = option.gasPrice || (await this.getGasPrice());
		const gasLimit = option.gasLimit || CST.DEFAULT_TX_GAS_LIMIT;
		const nonce = option.nonce || (await this.getTransactionCount(from));

		return this.web3.eth.sendTransaction({
			from: from,
			to: to,
			value: Web3Wrapper.toWei(value),
			gasPrice: gasPrice,
			gas: gasLimit,
			nonce: nonce
		});
	}

	public async erc20Transfer(
		contractAddress: string,
		from: string,
		to: string,
		value: number,
		option: ITransactionOption = {}
	): Promise<any> {
		if (this.isReadOnly()) return this.readOnlyReject();
		const gasPrice = option.gasPrice || (await this.getGasPrice());
		const gasLimit = option.gasLimit || CST.DEFAULT_TX_GAS_LIMIT;
		const nonce = option.nonce || (await this.getTransactionCount(from));
		const erc20Contract = this.createContract(erc20Abi.abi, contractAddress);

		return erc20Contract.methods.transfer(to, Web3Wrapper.toWei(value)).send({
			from: from,
			gasPrice: gasPrice,
			gas: gasLimit,
			nonce: nonce
		});
	}

	public async erc20Approve(
		contractAddress: string,
		from: string,
		spender: string,
		value: number,
		unlimited: boolean = false
	): Promise<any> {
		if (this.isReadOnly()) return this.readOnlyReject();
		return new Promise<string>(resolve => {
			const erc20Contract = this.createContract(erc20Abi.abi, contractAddress);
			return erc20Contract.methods
				.approve(
					spender,
					unlimited
						? Web3Wrapper.toHex(new BigNumber(2).pow(256).minus(1))
						: Web3Wrapper.toWei(value)
				)
				.send({
					from: from
				})
				.on('transactionHash', (txHash: string) => resolve(txHash));
		});
	}

	public getCurrentBlockNumber(): Promise<number> {
		return this.web3.eth.getBlockNumber();
	}

	public getBlock(
		blkNumber: number | string
	): Promise<{
		number: number;
		hash: string;
		parentHash: string;
		nonce: string;
		sha3Uncles: string;
		logsBloom: string;
		transactionsRoot: string;
		stateRoot: string;
		receiptsRoot: string;
		miner: string;
		difficulty: string;
		totalDifficulty: string;
		size: number;
		extraData: string;
		gasLimit: string;
		gasUsed: string;
		timestamp: number;
		transactions: string[];
		uncles: string[];
	}> {
		return this.web3.eth.getBlock(blkNumber);
	}

	public async getBlockTimestamp(blkNumber: number = 0) {
		const blk = await this.getBlock(blkNumber || 'latest');
		if (!blk)
			throw new Error('Error: no block found for ' + blkNumber);
		return blk.timestamp * 1000;
	}

	public async getCurrentAddress(): Promise<string> {
		const accounts = await this.web3.eth.getAccounts();
		return accounts[this.accountIndex] || CST.DUMMY_ADDR;
	}

	public getCurrentNetwork(): Promise<number> {
		return this.web3.eth.net.getId();
	}

	public async getEthBalance(address: string): Promise<number> {
		return Web3Wrapper.fromWei(await this.web3.eth.getBalance(address));
	}

	public async getErc20Balance(contractAddress: string, address: string) {
		const erc20Contract = this.createContract(erc20Abi.abi, contractAddress);

		return Web3Wrapper.fromWei(await erc20Contract.methods.balanceOf(address).call());
	}

	public async getErc20Allowance(contractAddress: string, address: string, spender: string) {
		const erc20Contract = this.createContract(erc20Abi.abi, contractAddress);

		return Web3Wrapper.fromWei(await erc20Contract.methods.allowance(address, spender).call());
	}

	public static fromWei(value: string | number) {
		return Number(Web3.utils.fromWei(value, 'ether'));
	}

	public static toWei(value: string | number): any {
		return Web3.utils.toWei(value + '', 'ether');
	}

	public static toHex(value: any): string {
		return Web3.utils.toHex(value);
	}

	public static checkAddress(addr: string): boolean {
		if (!addr.startsWith('0x') || addr.length !== 42) return false;
		return Web3.utils.checkAddressChecksum(Web3.utils.toChecksumAddress(addr));
	}

	public getTransactionReceipt(
		txHash: string
	): Promise<null | {
		status: boolean;
		blockHash: string;
		blockNumber: number;
		transactionHash: string;
		transactionIndex: number;
		from: string;
		to: string;
		contractAddress: string | null;
		cumulativeGasUsed: number;
		gasUsed: number;
		logs: Array<{
			data: string;
			topics: string[];
			logIndex: number;
			transactionIndex: number;
			transactionHash: string;
			blockHash: string;
			blockNumber: number;
			address: string;
		}>;
	}> {
		return this.web3.eth.getTransactionReceipt(txHash);
	}

	public static parseEvent(
		eventLog: {
			event: string;
			address: string;
			returnValues: any;
			logIndex: number;
			transactionIndex: number;
			transactionHash: string;
			blockHash: string;
			blockNumber: number;
			raw?: { data: string; topics: string[] };
		},
		timestamp: number
	): IEvent {
		const returnValue = eventLog.returnValues;
		const output: IEvent = {
			contractAddress: eventLog.address,
			type: eventLog.event,
			id: (eventLog as any)['id'],
			blockHash: eventLog.blockHash,
			blockNumber: eventLog.blockNumber,
			transactionHash: eventLog.transactionHash,
			logStatus: (eventLog as any)['type'],
			parameters: {},
			timestamp: timestamp
		};
		for (const key in returnValue) output.parameters[key] = returnValue[key];

		return output;
	}

	public static pullEvents(
		contract: any,
		start: number,
		end: number,
		event: string
	): Promise<
		Array<{
			event: string;
			address: string;
			returnValues: any;
			logIndex: number;
			transactionIndex: number;
			transactionHash: string;
			blockHash: string;
			blockNumber: number;
			raw?: { data: string; topics: string[] };
		}>
	> {
		return contract.getPastEvents(event, {
			fromBlock: start,
			toBlock: end
		});
	}

	public createContract(abi: any[], address: string): any {
		return new this.web3.eth.Contract(abi, address);
	}

	public getTransactionCount(address: string): Promise<number> {
		return this.web3.eth.getTransactionCount(address);
	}

	public async getTransactionOption(
		account: string,
		defaultGasLimit: number,
		option: ITransactionOption = {}
	) {
		const gasPrice = option.gasPrice || (await this.getGasPrice());
		const gasLimit = option.gasLimit || defaultGasLimit;
		const nonce = option.nonce || (await this.getTransactionCount(account));
		return {
			from: account,
			gasPrice: gasPrice,
			gas: gasLimit,
			nonce: nonce
		};
	}
}

export default Web3Wrapper;
