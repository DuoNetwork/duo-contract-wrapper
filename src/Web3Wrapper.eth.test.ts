// fix for @ledgerhq/hw-transport-u2f 4.28.0
import '@babel/polyfill';
import { Wallet } from './types';
import Web3Wrapper from './Web3Wrapper';

jest.mock('web3', () => {
	return jest.fn().mockImplementation(() => {
		return {
			eth: {
				getAccounts: jest.fn(() => Promise.resolve(['account1', 'account2', 'account3'])),
				getGasPrice: jest.fn(() => 1000000000),
				sendTransaction: jest.fn(),
				getTransactionCount: jest.fn(() => 10)
			}
		};
	});
});

test('getGasPrice', async () => {
	const web3Wrapper = new Web3Wrapper({ ethereum: {} }, '', '', true);
	expect(await web3Wrapper.getGasPrice()).toMatchSnapshot();
});

test('sendEther, without option', async () => {
	const web3Wrapper = new Web3Wrapper({ ethereum: {} }, '', '', true);
	web3Wrapper.toWei = jest.fn(() => '1000000000000000000');
	await web3Wrapper.sendEther('from', 'to', 1);
	expect((web3Wrapper.web3.eth.sendTransaction as jest.Mock).mock.calls).toMatchSnapshot();
});

test('switchToLedger', async () => {
	jest.mock('web3-provider-engine/subproviders/fetch');
	jest.mock('@ledgerhq/web3-subprovider', () => ({
		default: jest.fn()
	}));
	jest.mock('web3-provider-engine', () => {
		return jest.fn().mockImplementation(() => {
			return {
				addProvider: jest.fn(),
				start: jest.fn()
			};
		});
	});
	const web3Wrapper = new Web3Wrapper({ ethereum: {} }, '', '', true);
	web3Wrapper.handleSwitchToLedger = [jest.fn()];

	const res = await web3Wrapper.switchToLedger();
	expect(web3Wrapper.wallet).toBe(Wallet.Ledger);
	expect(res).toMatchSnapshot();
	expect(web3Wrapper.handleSwitchToLedger[0] as jest.Mock).toBeCalled();
});
