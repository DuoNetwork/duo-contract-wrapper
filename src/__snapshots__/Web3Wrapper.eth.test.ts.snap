// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`erc20Approve, readOnly 1`] = `"Read Only Mode"`;

exports[`erc20Approve, unlimited 1`] = `
Array [
  Array [
    "spender",
    "hex for 1.15792089237316195423570985008687907853269984665640564039457584007913129639935e+77",
  ],
]
`;

exports[`erc20Approve, unlimited 2`] = `
Array [
  Array [
    "spender",
    "hex for 1.15792089237316195423570985008687907853269984665640564039457584007913129639935e+77",
  ],
]
`;

exports[`erc20Approve, without option 1`] = `
Array [
  Array [
    "spender",
    "1000000000000000000",
  ],
]
`;

exports[`erc20Transfer, readOnly 1`] = `"Read Only Mode"`;

exports[`erc20Transfer, with option 1`] = `
Array [
  Array [
    "to",
    "1000000000000000000",
  ],
]
`;

exports[`erc20Transfer, without option 1`] = `
Array [
  Array [
    "to",
    "1000000000000000000",
  ],
]
`;

exports[`getBlock 1`] = `
Object {
  "blockHash": "blockHash",
  "timestamp": 123456789,
}
`;

exports[`getBlockTimestamp, no block 1`] = `[Error: Error: no block found for 0]`;

exports[`getBlockTimestamp, with blkNumber 1`] = `123456789000`;

exports[`getBlockTimestamp, without blkNumber 1`] = `123456789000`;

exports[`getCurrentAddress 1`] = `"0x0"`;

exports[`getCurrentBlockNumber 1`] = `100000`;

exports[`getCurretnNetwork 1`] = `42`;

exports[`getErc20Allowance 1`] = `0.001`;

exports[`getErc20Balance 1`] = `0.001`;

exports[`getEthBalance 1`] = `0.00001`;

exports[`getGasPrice 1`] = `1000000000`;

exports[`getTransactionReceipt 1`] = `"receipt"`;

exports[`sendEther, with option 1`] = `
Array [
  Array [
    Object {
      "from": "from",
      "gas": 20000,
      "gasPrice": 1000000000,
      "nonce": 10,
      "to": "to",
      "value": "1000000000000000000",
    },
  ],
]
`;

exports[`sendEther, without option 1`] = `
Array [
  Array [
    Object {
      "from": "from",
      "gas": 30000,
      "gasPrice": 1000000000,
      "nonce": 10,
      "to": "to",
      "value": "1000000000000000000",
    },
  ],
]
`;

exports[`switchToLedger 1`] = `
Array [
  "account1",
  "account2",
  "account3",
]
`;

exports[`switchToLedger kovan 1`] = `
Array [
  "account1",
  "account2",
  "account3",
]
`;
